src/utils.rs

```diff
@@ -67,50 +67,60 @@ impl<'de> Deserialize<'de> for CurvePoint {
 }
 
-/// Generate a random BigNumber in the range 0, ..., n
+/// Sample a number uniformly at random from the range [0, n). This can be used for sampling from
+/// a prime field `F_p` or the integers modulo `n` (for any `n`).
 pub(crate) fn random_positive_bn<R: RngCore + CryptoRng>(rng: &mut R, n: &BigNumber) -> BigNumber {
     BigNumber::from_rng(n, rng)
 }
 
-/// Generate a random BigNumber in the range -n, ..., n
-pub(crate) fn random_bn_plusminus<R: RngCore + CryptoRng>(rng: &mut R, n: &BigNumber) -> BigNumber {
-    let val = BigNumber::from_rng(n, rng);
+/// Sample a number uniformly at random from the range [-n, n].
+pub(crate) fn random_plusminus<R: RngCore + CryptoRng>(rng: &mut R, n: &BigNumber) -> BigNumber {
+    // `from_rng()` samples the _open_ interval, so add 1 to get the closed interval for `n`
+    let open_interval_max: BigNumber = n + 1;
+    let val = BigNumber::from_rng(&open_interval_max, rng);
     let is_positive: bool = rng.gen();
     match is_positive {
         true => val,
         false => -val,
     }
 }
 
-/// Generate a random BigNumber in the range -2^n, ..., 0, ..., 2^n
-pub(crate) fn random_bn_in_range<R: RngCore + CryptoRng>(rng: &mut R, n: usize) -> BigNumber {
-    let val = BigNumber::from_rng(&(BigNumber::one() << n), rng);
-    let is_positive: bool = rng.gen();
-    match is_positive {
-        true => val,
-        false => -val,
-    }
+/// Sample a number uniformly at random from the range `[-2^n, 2^n]`.
+pub(crate) fn random_plusminus_by_size<R: RngCore + CryptoRng>(rng: &mut R, n: usize) -> BigNumber {
+    let range = BigNumber::one() << n;
+    random_plusminus(rng, &range)
 }
 
-/// Generate a random BigNumber x in the range -2^n, ..., 0, ..., 2^n, where |x|
-/// > 2^min_bound
-#[cfg(test)]
-pub(crate) fn random_bn_in_range_min<R: RngCore + CryptoRng>(
+/// Sample a number uniformly at random from the range `[-scale * 2^n, scale * 2^n]`.
+pub(crate) fn random_plusminus_scaled<R: RngCore + CryptoRng>(
     rng: &mut R,
     n: usize,
-    min_bound: usize,
+    scale: &BigNumber,
+) -> BigNumber {
+    let range = (BigNumber::one() << n) * scale;
+    random_plusminus(rng, &range)
+}
+
+/// Sample a number uniformly at random from the range `[-2^max, -2^min] U [2^min, 2^max]`.
+#[cfg(test)]
+pub(crate) fn random_plusminus_by_size_with_minimum<R: RngCore + CryptoRng>(
+    rng: &mut R,
+    max: usize,
+    min: usize,
 ) -> crate::errors::Result<BigNumber> {
-    if min_bound >= n {
+    if min >= max {
         return bail!("min_bound needs to be less than n");
     }
-    let min_bound_bn = (BigNumber::one() << n) - (BigNumber::one() << min_bound);
-    let val = BigNumber::from_rng(&(min_bound_bn + (BigNumber::one() << min_bound)), rng);
+    // Sample from [0, 2^n - 2^min], then add 2^min to bump into correct range.
+    let min_bound_bn = (BigNumber::one() << max) - (BigNumber::one() << min);
+    let val = BigNumber::from_rng(&min_bound_bn, rng) + (BigNumber::one() << min);
```

Comment:
note: This is just used for testing but I intentionally changed the range here. We need to sample from the correct-size range, then bump the value up into the correctly-bounded range.